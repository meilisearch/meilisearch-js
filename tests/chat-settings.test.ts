import {
  afterAll,
  expect,
  test,
  describe,
  beforeEach,
  expectTypeOf,
} from "vitest";
import {
  ErrorStatusCode,
  type ChatSettings,
  type SearchParams,
} from "../src/types/index.js";
import {
  clearAllIndexes,
  config,
  getClient,
  dataset,
} from "./utils/meilisearch-test-utils.js";

const index = {
  uid: "chat_settings_test",
};

afterAll(() => {
  return clearAllIndexes(config);
});

describe.each([{ permission: "Master" }, { permission: "Admin" }])(
  "Test on chat settings",
  ({ permission }) => {
    beforeEach(async () => {
      await clearAllIndexes(config);
      const client = await getClient("Master");
      await client.index(index.uid).addDocuments(dataset).waitTask();
    });

    test(`${permission} key: Get an index chat settings`, async () => {
      const client = await getClient(permission);

      const response = await client.index(index.uid).getChat();
      expect(response).toHaveProperty("description");
      expect(response).toHaveProperty("documentTemplate");
      expect(response).toHaveProperty("documentTemplateMaxBytes");
      expect(response).toHaveProperty("searchParameters");
      expectTypeOf(response.searchParameters).toEqualTypeOf<SearchParams>();
    });

    test(`${permission} key: Update chat settings with all properties`, async () => {
      const client = await getClient(permission);
      const newChatSettings = {
        description:
          "A comprehensive movie database containing titles, overviews, genres, and release dates to help users find movies",
        documentTemplate:
          "Title: {{ title }}\nDescription: {{ overview }}\nGenres: {{ genres }}\nRelease Date: {{ release_date }}\n",
        documentTemplateMaxBytes: 500,
        searchParameters: {
          hybrid: { embedder: "default" },
          limit: 20,
          sort: ["release_date:desc"],
          distinct: "title",
          matchingStrategy: "last",
          attributesToSearchOn: ["title", "overview"],
          rankingScoreThreshold: 0.5,
        },
      } satisfies ChatSettings;

      await client.index(index.uid).updateChat(newChatSettings).waitTask();

      const response = await client.index(index.uid).getChat();
      expect(response).toMatchObject(newChatSettings);
    });
  },
);

describe.each([
  { permission: "Search", errorCode: ErrorStatusCode.INVALID_API_KEY },
  { permission: "No", errorCode: ErrorStatusCode.MISSING_AUTHORIZATION_HEADER },
])("Test on chat settings", ({ permission, errorCode }) => {
  beforeEach(async () => {
    await clearAllIndexes(config);
    const client = await getClient("Master");
    await client.createIndex(index.uid).waitTask();
  });

  test(`${permission} key: try to get chat settings and be denied`, async () => {
    const client = await getClient(permission);
    await expect(client.index(index.uid).getChat()).rejects.toHaveProperty(
      "cause.code",
      errorCode,
    );
  });

  test(`${permission} key: try to update chat settings and be denied`, async () => {
    const client = await getClient(permission);
    await expect(
      client.index(index.uid).updateChat({ description: "test" }),
    ).rejects.toHaveProperty("cause.code", errorCode);
  });
});
