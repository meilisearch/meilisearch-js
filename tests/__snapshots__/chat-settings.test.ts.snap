// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test on chat settings > Admin key: Get an index chat settings 1`] = `
{
  "description": "",
  "documentTemplate": "{% for field in fields %}{% if field.is_searchable and field.value != nil %}{{ field.name }}: {{ field.value }}
{% endif %}{% endfor %}",
  "documentTemplateMaxBytes": 400,
  "searchParameters": {},
}
`;

exports[`Test on chat settings > Admin key: Get default chat settings on index 1`] = `
{
  "description": "",
  "documentTemplate": "{% for field in fields %}{% if field.is_searchable and field.value != nil %}{{ field.name }}: {{ field.value }}
{% endif %}{% endfor %}",
  "documentTemplateMaxBytes": 400,
  "searchParameters": {},
}
`;

exports[`Test on chat settings > Admin key: Update chat settings with all properties 1`] = `
{
  "description": "A comprehensive movie database containing titles, overviews, genres, and release dates to help users find movies",
  "documentTemplate": "Title: {{ title }}
Description: {{ overview }}
Genres: {{ genres }}
Release Date: {{ release_date }}
",
  "documentTemplateMaxBytes": 500,
  "searchParameters": {
    "attributesToSearchOn": [
      "title",
      "overview",
    ],
    "distinct": "title",
    "hybrid": {
      "embedder": "default",
      "semanticRatio": 0.5,
    },
    "limit": 20,
    "matchingStrategy": "last",
    "rankingScoreThreshold": 0.5,
    "sort": [
      "release_date:desc",
    ],
  },
}
`;

exports[`Test on chat settings > Master key: Get an index chat settings 1`] = `
{
  "description": "",
  "documentTemplate": "{% for field in fields %}{% if field.is_searchable and field.value != nil %}{{ field.name }}: {{ field.value }}
{% endif %}{% endfor %}",
  "documentTemplateMaxBytes": 400,
  "searchParameters": {},
}
`;

exports[`Test on chat settings > Master key: Get default chat settings on index 1`] = `
{
  "description": "",
  "documentTemplate": "{% for field in fields %}{% if field.is_searchable and field.value != nil %}{{ field.name }}: {{ field.value }}
{% endif %}{% endfor %}",
  "documentTemplateMaxBytes": 400,
  "searchParameters": {},
}
`;

exports[`Test on chat settings > Master key: Update chat settings with all properties 1`] = `
{
  "description": "A comprehensive movie database containing titles, overviews, genres, and release dates to help users find movies",
  "documentTemplate": "Title: {{ title }}
Description: {{ overview }}
Genres: {{ genres }}
Release Date: {{ release_date }}
",
  "documentTemplateMaxBytes": 500,
  "searchParameters": {
    "attributesToSearchOn": [
      "title",
      "overview",
    ],
    "distinct": "title",
    "hybrid": {
      "embedder": "default",
      "semanticRatio": 0.5,
    },
    "limit": 20,
    "matchingStrategy": "last",
    "rankingScoreThreshold": 0.5,
    "sort": [
      "release_date:desc",
    ],
  },
}
`;
